### PR TITLE
[Feature][CDC][Zeta] Support shuffle multiple rows by tableId

### LINKED ISSUE #4054: [Feature][Zeta] Make SeaTunnel Zeta support multi-table
### Search before asking

- [X] I had searched in the [feature](https://github.com/apache/incubator-seatunnel/issues?q=is%3Aissue+label%3A%22Feature%22) and found no similar feature requirement.

### Description

This is a subtask of https://github.com/apache/incubator-seatunnel/issues/4050 to track completion.

Make the Zeta engine support multi-table data distribution to the corresponding data channel.

#### Related pseudo-code

```java
// For Source deserialization and Row distribution
public class MultipleRowType implements SeaTunnelDataType<SeaTunnelRow> {
    private final String[] tableNames;
    private final SeaTunnelRowType[] rowTypes;
}
```

```java
// pseudo-code
public class DistributionTransform extends SeaTunnelTransform<Record<?>> {
    // Use MultipleRowType to distribute records to corresponding data channels
    private MultipleRowType multiRowType;
}
```

### Usage Scenario

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by hailin0**
## Purpose of this pull request

close #4054

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by hailin0** [comment]
@ashulin @Hisoka-X PTAL

File: seatunnel-connectors-v2/connector-cdc/connector-cdc-base/src/main/java/org/apache/seatunnel/connectors/cdc/debezium/row/SeaTunnelRowDebeziumDeserializeSchema.java [hunk 1]
```diff
@@ -114,34 +113,32 @@ public void deserialize(SourceRecord record, Collector<SeaTunnelRow> collector)
         Schema valueSchema = record.valueSchema();
 
         Struct sourceStruct = messageStruct.getStruct(Envelope.FieldName.SOURCE);
-        String database = sourceStruct.getString(DATABASE_NAME_KEY);
         String tableName = sourceStruct.getString(AbstractSourceInfo.TABLE_NAME_KEY);
-        String tableId = database + ":" + tableName;
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-cdc/connector-cdc-base/src/main/java/org/apache/seatunnel/connectors/cdc/debezium/row/SeaTunnelRowDebeziumDeserializeSchema.java:119]: Why change tableId to tableName not database+tableName, shouldn't zeta support multi database in one pipeline?

File: seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/actions/ShuffleConfig.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+            IQueue<Record<?>>[] shuffles = new IQueue[getInputPartitions()];
+            for (int inputIndex = 0; inputIndex < getInputPartitions(); inputIndex++) {
+                String queueName = generateQueueName(inputIndex, targetIndex);
+                shuffles[inputIndex] = getIQueue(hazelcast, queueName);
+            }
+            return shuffles;
+        }
+
+        private String generateQueueName(int inputIndex, int targetIndex) {
+            return String.format("PartitionShuffle-Queue[%s-%s-%s]", getJobId(), inputIndex, targetIndex);
```
Comment (Hisoka-X) [inline seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/actions/ShuffleConfig.java]: Should use jobId and PipelineId

File: seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/actions/ShuffleConfig.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+            IQueue<Record<?>>[] queues = new IQueue[getInputPartitions()];
+            for (int inputIndex = 0; inputIndex < getInputPartitions(); inputIndex++) {
+                String queueName = generateQueueName(inputIndex, targetTableId);
+                queues[inputIndex] = getIQueue(hazelcast, queueName);
+            }
+            return queues;
+        }
+
+        private String generateQueueName(int inputIndex, String tableId) {
+            return "MultipleRowShuffle-Queue[" + getJobId() + "-" + inputIndex + "-" + tableId + "]";
```
Comment (Hisoka-X) [inline seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/actions/ShuffleConfig.java]: Same with above

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/master/JobMaster.java [hunk 4]
```diff
@@ -174,40 +176,29 @@ public void init(long initializationTimestamp) {
...
 
     // TODO replace it after ReadableConfig Support parse yaml format, then use only one config to read engine and env config.
-    private CheckpointConfig mergeEnvAndEngineConfig(CheckpointConfig engine, Map<String, Object> env) {
-        CheckpointConfig checkpointConfig = new CheckpointConfig();
+    private CheckpointConfig mergeEnvAndEngineConfig(CheckpointConfig engineConfig, Map<String, Object> env) {
         if (env.containsKey(EnvCommonOptions.CHECKPOINT_INTERVAL.key())) {
-            checkpointConfig.setCheckpointInterval((Integer) env.get(EnvCommonOptions.CHECKPOINT_INTERVAL.key()));
+            engineConfig.setCheckpointInterval((Integer) env.get(EnvCommonOptions.CHECKPOINT_INTERVAL.key()));
         }
-        checkpointConfig.setCheckpointTimeout(engine.getCheckpointTimeout());
```
Comment (Hisoka-X) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/master/JobMaster.java:202]: Why remove checkpoint config for timeout and others?

File: seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/actions/ShuffleConfig.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        }
+
+        @Override
+        public Map<String, IQueue<Record<?>>> createShuffles(HazelcastInstance hazelcast, int inputIndex) {
+            checkArgument(inputIndex >= 0 && inputIndex < getInputPartitions());
+            Map<String, IQueue<Record<?>>> shuffleMap = new HashMap<>();
+            for (int targetIndex = 0; targetIndex < getTargetPartitions(); targetIndex++) {
+                String queueName = generateQueueName(inputIndex, targetIndex);
+                IQueue<Record<?>> queue = getIQueue(hazelcast, queueName);
+                shuffleMap.put(queueName, queue);
```
Comment (Hisoka-X) [inline seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/actions/ShuffleConfig.java]: Maybe should call `clear` before put it into map

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/dag/physical/PhysicalPlanGenerator.java [hunk 6]
```diff
@@ -463,13 +462,6 @@ private void setFlowConfig(Flow f, int parallelismIndex) {
                     config.setCommitterTask(committerTaskIDMap.get((SinkAction<?, ?, ?, ?>) flow.getAction()));
                 }
                 flow.setConfig(config);
-            } else if (flow.getAction() instanceof ShuffleAction) {
```
Comment (Hisoka-X) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/dag/physical/PhysicalPlanGenerator.java:466]: Where parse executionPlan about shuffle logic to phyiscalPlan ?

File: seatunnel-connectors-v2/connector-cdc/connector-cdc-base/src/main/java/org/apache/seatunnel/connectors/cdc/debezium/row/SeaTunnelRowDebeziumDeserializeSchema.java (hunk 1 above)
Comment (hailin0) [reply seatunnel-connectors-v2/connector-cdc/connector-cdc-base/src/main/java/org/apache/seatunnel/connectors/cdc/debezium/row/SeaTunnelRowDebeziumDeserializeSchema.java:119]: No need to support multiple databases for now

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/master/JobMaster.java (hunk 4 above)
Comment (hailin0) [reply seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/master/JobMaster.java:202]: fixed

File: seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/actions/ShuffleConfig.java (hunk 3 above)
Comment (hailin0) [reply seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/actions/ShuffleConfig.java]: fixed

File: seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/actions/ShuffleConfig.java (hunk 2 above)
Comment (hailin0) [reply seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/actions/ShuffleConfig.java]: fixed

File: seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/actions/ShuffleConfig.java (hunk 5 above)
Comment (hailin0) [reply seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/actions/ShuffleConfig.java]: fixed

**Review by Hisoka-X — APPROVED** [review]
LGTM

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/dag/execution/ExecutionPlanGenerator.java [hunk 7]
```diff
@@ -88,17 +96,79 @@ public class ExecutionPlanGenerator {
...
+                            .setMaxSize(ShuffleConfig.DEFAULT_QUEUE_SIZE)
+                            .setBackupCount(ShuffleConfig.DEFAULT_QUEUE_BACKUP_COUNT)
+                            .setAsyncBackupCount(ShuffleConfig.DEFAULT_QUEUE_ASYNC_BACKUP_COUNT)
+                            .setEmptyQueueTtl((int) (checkpointConfig.getCheckpointInterval() * 3)));
+                    ShuffleConfig shuffleConfig = new ShuffleConfig()
+                        .setBatchSize(ShuffleConfig.DEFAULT_BATCH_SIZE)
+                        .setBatchFlushInterval(ShuffleConfig.DEFAULT_BATCH_FLUSH_INTERVAL)
+                        .setShuffleStrategy(shuffleStrategy);
+
+                    long shuffleVertexId = idGenerator.getNextId();
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/dag/execution/ExecutionPlanGenerator.java]: Generate duplicate id, waiting for fix

File: pom.xml [hunk 8]
```diff
@@ -142,7 +142,7 @@
         <javax.servlet.jap.version>2.1</javax.servlet.jap.version>
         <hadoop.binary.version>2.7</hadoop.binary.version>
         <jackson.version>2.12.6</jackson.version>
-        <lombok.version>1.18.0</lombok.version>
+        <lombok.version>1.18.24</lombok.version>
```
Comment (hailin0) [inline pom.xml:145]: upgrade use `@SuperBuilder`

File: seatunnel-connectors-v2/connector-cdc/connector-cdc-base/src/main/java/org/apache/seatunnel/connectors/cdc/debezium/row/SeaTunnelRowDebeziumDeserializeSchema.java [hunk 9]
```diff
@@ -114,34 +113,32 @@ public void deserialize(SourceRecord record, Collector<SeaTunnelRow> collector)
...
-        SeaTunnelRowDebeziumDeserializationConverters converters = multipleRowConverters.getOrDefault(tableId, singleRowConverter);
+        SeaTunnelRowDebeziumDeserializationConverters converters = multipleRowConverters.getOrDefault(tableName, singleRowConverter);
 
         if (operation == Envelope.Operation.CREATE || operation == Envelope.Operation.READ) {
             SeaTunnelRow insert = extractAfterRow(converters, record, messageStruct, valueSchema);
             insert.setRowKind(RowKind.INSERT);
-            insert.setTableId(tableId);
+            insert.setTableId(tableName);
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-cdc/connector-cdc-base/src/main/java/org/apache/seatunnel/connectors/cdc/debezium/row/SeaTunnelRowDebeziumDeserializeSchema.java:122]: SeaTunnelRow#tableId = tableName

File: seatunnel-connectors-v2/connector-console/src/main/java/org/apache/seatunnel/connectors/seatunnel/console/sink/ConsoleSinkWriter.java [hunk 10]
```diff
@@ -36,7 +36,7 @@
 public class ConsoleSinkWriter extends AbstractSinkWriter<SeaTunnelRow, Void> {
 
     private final SeaTunnelRowType seaTunnelRowType;
-    public static final AtomicLong CNT = new AtomicLong(0);
+    public final AtomicLong CNT = new AtomicLong(0);
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-console/src/main/java/org/apache/seatunnel/connectors/seatunnel/console/sink/ConsoleSinkWriter.java]: Fix the abnormal statistics of parallel sink

**Review by ashulin — APPROVED** [review]
LGTM

**Pull request merged**